File: week5/n-ary-tree-postorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> postorder(Node root) {
        List<Integer> list = new ArrayList<>();
        postorder(root, list);
        return list;
    }
    
    public void postorder(Node root, List<Integer> list) {
        if (root == null) {
            return;
        }
        if (root.children.size() == 0) {
            list.add(root.val);
            return;
        }
        for (Node node : root.children) {
            postorder(node, list);
        }
        list.add(root.val);
    }
}

File: summer2021/unit5/n-ary-tree-postorder-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        result = []
        self.helper(root, result)
        return result
    
    def helper(self, root, result):
        if not root:
            return
        for child in root.children:
            self.helper(child, result)
        result.append(root.val)
            
        